### DISCUSSION TITLE
ParseException when MERGE is used in a statement

### DISCUSSION THREAD

**Discussion by badapinguino** [JSQLParser/JSqlParser#1862] [Q&A] [answered]
Hello everyone,

I'm trying to parse a statement to get the target and source tables and save them into a file.
I already did it successfully with INSERT and UPDATE, also when there are joins, unions, subqueries, and WITHs. 
Unfortunately, I'm having issues with the MERGE clause.
For example, I have the following MERGE statement:

```
WITH
        WMachine AS
        (   SELECT
                DISTINCT 
                ProjCode,
                PlantCode,
                BuildingCode,
                FloorCode,
                Room
            FROM
                TAB_MachineLocation
            WHERE
                TRIM(Room) <> '' AND TRIM(Room) <> '-'
        )
    MERGE
        TAB_RoomLocation AS TRoom
    USING
        WMachine
    ON
        (
            TRoom.ProjCode = WMachine.ProjCode
        AND TRoom.PlantCode = WMachine.PlantCode
        AND TRoom.BuildingCode = WMachine.BuildingCode
        AND TRoom.FloorCode = WMachine.FloorCode
        AND TRoom.Room = WMachine.Room)
    WHEN NOT MATCHED BY TARGET THEN
    INSERT
        (
            ProjCode,
            PlantCode,
            BuildingCode,
			FloorCode,
            Room
        )
        VALUES
        (
            WMachine.ProjCode,
            WMachine.PlantCode,
            WMachine.BuildingCode,
            WMachine.FloorCode,
            WMachine.Room
        )
        OUTPUT GETDATE() AS TimeAction,
        $action as Action,
        INSERTED.ProjCode,
        INSERTED.PlantCode,
        INSERTED.BuildingCode,
        INSERTED.FloorCode,
        INSERTED.Room
    INTO
        TAB_MergeActions_RoomLocation
```

It works fine on SQL Server but when I try to parse it with jsqlparser with the function _CCJSqlParserUtil.parse(statement)_  it throws a net.sf.jsqlparser.parser.ParseException, that says: 
```
Encountered unexpected token: "TAB_RoomLocation" <S_IDENTIFIER>
    at line 16, column 9.

Was expecting:

    "INTO"
```

Why does jsqlparser expect an INTO after the MERGE keyword? My SQL code works fine on SQL Server.

Thank you in advance for your help.

**Comment by manticore-projects** [comment]
Greetings,

your syntax is not SQL:2016 compliant.
Please see the supported synatx for the `Merge Statement` here: http://217.160.215.75:8080/jsqlformatter/JSQLParser/syntax_snapshot.html#merge

**Reply by manticore-projects** [reply]
You can also [test your statement online here.](http://jsqlformatter.manticore-projects.com/jsqlformatter/demo.html?args=-c%20OoSQKgEgUABH92AWQIYGMAWBLAdgUxgEEBlWBeACnmIFEAZGgYTDPLfgBERiwQA5ZjFbs2ABQBOAewBWjSQBM8AGmEiEogDYocAFzmKVa9gCEArlg3zcAc33LVRmADENkyeLuHHCAEpuAtg5sTj4A8khBamCExgD6qJi4eHSSaCg6WJI4kQjAEDQ+NDnsYD4gSBR+kv4AlDAAPAB8MADkLUR8HDCl5ZUBdU2tALQtkTXCSAUA4kUi0XFV-ilpGVlExN2LwgCqxPxTkcjo2PjCoXyRFMVwpQEAdBIydjAAvDBHifgPUrIKeJGETqbe6abR6P6vd4JE54B5aXR2AFA27VO5mCxWHC2CFvD4wtHmSw2REiQFdFH+O4uNweHFQ45JKmudwk9hk4GoxaQvGMxbjXL5PgwPihMAwJCEMCMfJdYwATW6hB8MzFkBoFwQ-FoPhYIiu3kevwM1xgoIRfy8RnRRKxnhN1JZFpNWxE-PYADVCHRtjRSHqTTyvoa7d5A7CzeDjaHoYzrZjsVHHGGmTSQ0mY18Xew3WxQtswKJ8zAVRxJTQKHUSN0sP48IQ0KscJbyAASdCNmAoADORAbmSbkS1BTANA43yeTpEQ51o7hYLTbGnI7HceJk-YS9nDtpicXfG1y7uWbg-DAoUi83ieHE1jrfayXdii2W6X7QA)

**Comment by manticore-projects** [comment]
You SQL:2016 compliant `Merge` would look like this:

```sql
MERGE INTO tab_roomlocation AS troom
    USING wmachine
        ON ( troom.projcode = wmachine.projcode
                AND troom.plantcode = wmachine.plantcode
                AND troom.buildingcode = wmachine.buildingcode
                AND troom.floorcode = wmachine.floorcode
                AND troom.room = wmachine.room )
WHEN NOT MATCHED THEN
    INSERT ( projcode
                , plantcode
                , buildingcode
                , floorcode
                , room )
    VALUES ( wmachine.projcode
                , wmachine.plantcode
                , wmachine.buildingcode
                , wmachine.floorcode
                , wmachine.room )
;
```

We do not support the `Output` clause for the `MergeInsert`.

**Comment by badapinguino** [comment]
Hello,

Thank you for your answers and the code example.

I tested the code with the link you sent me and it shows me an error, as you said.
It's strange that it's unsupported because I ran this code biweekly in a stored procedure on SQL Server 2019, and it never showed me any error and it works as expected.
Maybe it's a syntax supported only in SQL Server 2019 and not in 2016?
Anyway thank you for specifying that the OUTPUT clause is not supported, it's a pity but I will keep that in mind.

I will configure my program to accept the Merge statements with the syntax you showed me.

Have a nice day.

**Comment by manticore-projects** [comment, marked as answer]
Greetings!

I have added support for `WITH ...` and the `Output` Clause.
[You can test your statement online here.](http://jsqlformatter.manticore-projects.com/jsqlformatter/demo.html?args=-c%20OoSQKgEgBA7gtgQwMYAsCWA7AplAggZSgAoAoKci8-AUQBlqBhMKAERHzBADkmzL-yABwBOAewBWSUQBMsfAZQA0UQQBsEGAC5TZ8heWUAjAK5pV0zAHMdc-UqgAzVaNHCbe-crGi4HigDEAJQB5AFkoTQRDAH1EVEwsZyQETTRRDD9yYAhqQOooMGE0OCIoQJc4KABKKAAeAD4oAHImzP5cLhYCopKyiuq6xqaAWibqklDcgHF87jBgiKjo7zgklLSMPEJNFb0AVXxuKdg49Gw24K5iCJWAOhEJGygAXhPkM6x7sUkZWzsBDpdHYVe7qLRPV7wd4JUEabS-Nr6QE3EEmMwWDDWX4vN7xbC3NHmKzuf4Azoony3JwuNzYyGnGHU1wk0mUZHAykrHFQvGfLlVEjZahXLjBZihXBgBg5LoAegAVFAAEIATQKuECM2Y8tlBRyXD03BogWYpQeP10rPsajhLKtUCMpiJmLtVuUTNplvtDqg-L0ADVcLQ9tRCKUeR8vo8Ed6fRGYTbwTHvcp4-jCRisV77amGfiPa7Wbnofi-cE9mAAAoV8hTLCaaQpLBEGoECLFLDIVLpNrKAAkXY2Wygg57nigmAAzlhhJosNIoxa-gJlFOZ3OF4n4dn+KuMNPZ-OCU7M4WKHuDxuqc5mcmVxP9+uj7tKHMFpEYnAZ5ZO0hu-vlgqNZ-xIABuIA)

```sql
WITH wmachine AS (
        SELECT DISTINCT
            projcode
            , plantcode
            , buildingcode
            , floorcode
            , room
        FROM tab_machinelocation
        WHERE Trim( Room ) <> ''
            AND Trim( Room ) <> '-' )
MERGE INTO tab_roomlocation AS troom
    USING wmachine
        ON ( troom.projcode = wmachine.projcode
                AND troom.plantcode = wmachine.plantcode
                AND troom.buildingcode = wmachine.buildingcode
                AND troom.floorcode = wmachine.floorcode
                AND troom.room = wmachine.room )
WHEN NOT MATCHED /* BY TARGET */ THEN
    INSERT ( projcode
                , plantcode
                , buildingcode
                , floorcode
                , room )
    VALUES ( wmachine.projcode
                , wmachine.plantcode
                , wmachine.buildingcode
                , wmachine.floorcode
                , wmachine.room )
OUTPUT  Getdate() AS timeaction
        , $action AS action
        , inserted.projcode
        , inserted.plantcode
        , inserted.buildingcode
        , inserted.floorcode
        , inserted.room
    INTO tab_mergeactions_roomlocation
;
```

**Reply by badapinguino** [reply]
Wow, that's great! Thank you so much.

I will test it when I finish implementing the MERGE management in my program.
And I will change the MERGE syntax used in my store procedure to the one supported by you.

Thank you again and have a great day.
